pkg/apis/monitoring/v1/types.go

```diff
@@ -1042,6 +1046,44 @@ type ProberSpec struct {
 	ProxyURL string `json:"proxyUrl,omitempty"`
 }
 
+// OAuth2 allows an endpoint to authenticate with OAuth2.
+// More info: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#oauth2
+// +k8s:openapi-gen=true
+type OAuth2 struct {
+	// The secret or configmap containing the OAuth2 client id
+	ClientID SecretOrConfigMap `json:"clientId"`
+	// The secret containing the OAuth2 client secret
+	ClientSecret v1.SecretKeySelector `json:"clientSecret"`
+	// The URL to fetch the token from
+	TokenURL string `json:"tokenUrl"`
+	// OAuth2 scopes used for the token request
+	Scopes []string `json:"scopes,omitempty"`
+	// Parameters to append to the token URL
+	EndpointParams map[string]string `json:"endpointParams,omitempty"`
+}
+
+type OAuth2ValidationError struct {
+	err string
+}
+
+func (e *OAuth2ValidationError) Error() string {
+	return e.err
+}
+
+func (o *OAuth2) Validate() error {
```

Comment:
shouldn't we check that `TokenURL` isn't empty too?

Reply:
I think that the TokenURL cannot be empty because it is marked as required here: https://github.com/prometheus-operator/prometheus-operator/pull/4113/files#diff-73c670418fb192fb77d3a1c95ada8448ee256f1a6afa0e6ede380a7c63221184R12025
But I don't mind adding it for consistency reasons if you think we should do it. cc @jan--f

Reply:
I think that an empty string value would pass the required validation but worth checking. We can add `// +kubebuilder:validation:MinLength=1` before the field declaration.
An additional check in `Validate()` wouldn't hurt because you might be able to bypass the server-side validation (I couldn't the exact option in kubectl though).

Reply:
Good points, I wasn't aware of these gaps. I applied both suggestions, adding the OpenAPI minLength validation and extending the `Validate()` method for `OAuth2` to return an error if the token is empty.